internal/controllers/openshift.go

```diff
@@ -0,0 +1,199 @@
+		return nil
+	}
+	err := r.reconcileConsoleLink(ctx, cr)
+	if err != nil {
+		return err
+	}
+	return r.addCorsAllowedOriginIfNotPresent(ctx, cr)
+}
+
+func (r *CryostatReconciler) finalizeOpenShift(ctx context.Context, cr *operatorv1beta1.Cryostat) error {
+	if !r.IsOpenShift {
+		return nil
+	}
+	err := r.deleteConsoleLink(ctx, newConsoleLink(cr))
+	if err != nil {
+		return err
+	}
+	return r.deleteCorsAllowedOrigins(ctx, cr)
+}
+
+func newConsoleLink(cr *operatorv1beta1.Cryostat) *consolev1.ConsoleLink {
+	// Cluster scoped, so use a unique name to avoid conflicts
+	return &consolev1.ConsoleLink{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: common.ClusterUniqueName(cr),
+		},
+	}
+}
+
+func (r *CryostatReconciler) reconcileConsoleLink(ctx context.Context, cr *operatorv1beta1.Cryostat) error {
+	link := newConsoleLink(cr)
+
+	url := cr.Status.ApplicationURL
+	if len(url) == 0 {
+		// Nothing to link to, so remove the link if it exists
+		err := r.deleteConsoleLink(ctx, link)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	op, err := controllerutil.CreateOrUpdate(ctx, r.Client, link, func() error {
+		link.Spec.Link = consolev1.Link{
+			Text: "Cryostat",
+			Href: url,
+		}
+		link.Spec.Location = consolev1.NamespaceDashboard
+		link.Spec.NamespaceDashboard = &consolev1.NamespaceDashboardSpec{
+			Namespaces: []string{cr.Namespace},
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	r.Log.Info(fmt.Sprintf("Console Link %s", op), "name", link.Name)
+	return nil
+}
+
+func (r *CryostatReconciler) deleteConsoleLink(ctx context.Context, link *consolev1.ConsoleLink) error {
+	err := r.Client.Delete(ctx, link)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			r.Log.Info("ConsoleLink not found, proceeding with deletion", "name", link.Name)
+			return nil
+		}
+		r.Log.Error(err, "failed to delete ConsoleLink", "name", link.Name)
+		return err
+	}
+	r.Log.Info("deleted ConsoleLink", "name", link.Name)
+	return nil
+}
+
+func (r *CryostatReconciler) addCorsAllowedOriginIfNotPresent(ctx context.Context, cr *operatorv1beta1.Cryostat) error {
+	reqLogger := r.Log.WithValues("Request.Namespace", cr.Namespace, "Request.Name", cr.Name)
+
+	allowedOrigin := cr.Status.ApplicationURL
+	if len(allowedOrigin) == 0 {
+		return nil
+	}
+
+	apiServer := &configv1.APIServer{}
+	err := r.Client.Get(ctx, types.NamespacedName{Name: apiServerName}, apiServer)
+	if err != nil {
+		reqLogger.Error(err, "Failed to get APIServer config")
+		return err
+	}
+
+	allowedOriginAsRegex := regexp.QuoteMeta(allowedOrigin)
+
+	for _, origin := range apiServer.Spec.AdditionalCORSAllowedOrigins {
+		if origin == allowedOriginAsRegex {
+			return nil
+		}
+	}
+
+	apiServer.Spec.AdditionalCORSAllowedOrigins = append(
+		apiServer.Spec.AdditionalCORSAllowedOrigins,
+		allowedOriginAsRegex,
+	)
+
+	err = r.Client.Update(ctx, apiServer)
+	if err != nil {
+		reqLogger.Error(err, "Failed to update APIServer CORS allowed origins")
+		return err
+	}
+
+	return nil
```

Comment:
Should have some logs for this operations? Same for `deleteCorsAllowedOrigins`?

Reply:
Good idea. I've updated the logging for this file.